src/python/swarm-lifecycle-event-handler/app.py

```diff
@@ -0,0 +1,174 @@
+        AutoScalingGroupName=autoscaling_group_name,
+        InstanceId=instance_id,
+        LifecycleActionResult="CONTINUE",
+    )
+    LOGGER.info(
+        f"Completed {lifecycle_hook_name} lifecycle action for instance {instance_id} in ASG {autoscaling_group_name}"
+    )
+
+
+def _remove_dns_ip(dns_name: str, ip_address: str, hosted_zone_id: str) -> None:
+    route53 = boto3.client("route53")
+    ip_addresses = [
+        ip
+        for ip in _dns_ip_addresses(route53, dns_name, None, hosted_zone_id)
+        if ip != ip_address
+    ]
+
+    change = {
+        "Action": "DELETE",  # delete the A record if this is the last address
+        "ResourceRecordSet": {
+            "Name": dns_name,
+            "Type": "A",
+            "TTL": 300,
+            "ResourceRecords": [{"Value": ip_address}],
+        },
+    }
+    if len(ip_addresses) > 0:
+        change["Action"] = "UPSERT"
+        change["ResourceRecordSet"]["ResourceRecords"] = [
+            {"Value": ip} for ip in ip_addresses
+        ]
+
+    try:
+        comment = f"Removed {ip_address} from {dns_name} DNS A Record"
+        route53.change_resource_record_sets(
+            HostedZoneId=hosted_zone_id,
+            ChangeBatch={
+                "Changes": [change],
+                "Comment": comment,
+            },
+        )
+        LOGGER.info(comment)
+    except ClientError as e:
+        if e.response["Error"]["Code"] == "InvalidChangeBatch":
+            LOGGER.warn(f"DNS record does not exist for {ip_address}")
+        else:
+            raise e
+
+
+def _insert_dns_ip(dns_name: str, ip_address: str, hosted_zone_id: str) -> None:
+    route53 = boto3.client("route53")
+    comment = f"Inserted {ip_address} into {dns_name} DNS A Record"
+    route53.change_resource_record_sets(
+        HostedZoneId=hosted_zone_id,
+        ChangeBatch={
+            "Changes": [
+                {
+                    "Action": "UPSERT",
+                    "ResourceRecordSet": {
+                        "Name": dns_name,
+                        "Type": "A",
+                        "TTL": 300,
+                        "ResourceRecords": [
+                            {"Value": ip}
+                            for ip in _dns_ip_addresses(
+                                route53, dns_name, ip_address, hosted_zone_id
+                            )
+                        ],
+                    },
+                },
+            ],
+            "Comment": comment,
+        },
+    )
+    LOGGER.info(comment)
+
+
+@app.lambda_function()
+def main(event, context) -> None:
+    LOGGER.debug(f"Processing event: {json.dumps(event)}")
+    for record in event["Records"]:
+        LOGGER.debug(f"Processing record: {json.dumps(record)}")
+        if "Sns" in record:
+            message = json.loads(record["Sns"]["Message"])
+            transition = message["LifecycleTransition"]
+
+            if "NotificationMetadata" in message:
```

Comment:
don't super love the duck typing based on "does this substring appear in json" tbh
i think for a lot of these you can just do like
```
notif_meta = message.get("NotificationMetadata", None)
if notif_meta:
```